### ISSUE TITLE
rotate180_rgb565 code optimization to yield >= 25% increase in speed

### ISSUE THREAD

**Issue by kdschlosser** [kisvegabor/lvgl#6827]
### Introduce the problem

Here is the existing function that is located in `src/draw/sw/lv_draw_sw.c`

```c
static void rotate180_rgb565(const uint16_t * src, uint16_t * dst, int32_t width, int32_t height, int32_t src_stride,
                             int32_t dest_stride)
{
    if(LV_RESULT_OK == LV_DRAW_SW_ROTATE180_RGB565(src, dst, srcWidth, srcHeight, srcStride, dstStride)) {
        return ;
    }

    src_stride /= sizeof(uint16_t);
    dest_stride /= sizeof(uint16_t);

    for(int32_t y = 0; y < height; ++y) {
        int32_t dstIndex = (height - y - 1) * dest_stride;
        int32_t srcIndex = y * src_stride;
        for(int32_t x = 0; x < width; ++x) {
            dst[dstIndex + width - x - 1] = src[srcIndex + x];
        }
    }
}
```

Now the first thing is I am not even sure how this is even managing to compile without pitching a fit about the `srcWidth`, `srcHeight`, `srcStride` and `dstStride` variables not existing. Those are passed as parameters to the `LV_DRAW_SW_ROTATE180_RGB565` macro.

But never mind that at the moment. I am not sure as to why there is all of this crazy math that is being used in order to rotate the display 180°. This can be simplified using the code below

### Proposal

```c
static void rotate180_rgb565(const uint16_t * src, uint16_t * dst, int32_t width, int32_t height, int32_t src_stride,
                             int32_t dest_stride)
{
    if(LV_RESULT_OK == LV_DRAW_SW_ROTATE180_RGB565(src, dst, srcWidth, srcHeight, srcStride, dstStride)) {
        return ;
    }
    // not sure if these need to be here or not. I believe it might matter if the macro above is not defined in userland
    LV_UNUSED(srcStride);
    LV_UNUSED(dstStride);

    uint32_t num_pixels = height * width;

    for(int32_t i = 0; i < num_pixels; i++) {
        dst[~i + num_pixels] = src[i];
    }
}
```

The code above yields a 25% increase in speed at a minimum. The results of the speed test are with the code converted to python running the test 5000 times over each function. I believe the `~` operator does the same in C code that it does in Python. In Python `~0` is `-1`.

so basically we are taking a  function that has 2 division and 2 addition, 4 subtraction, 2 multiplication done in for loops and reducing it to a bit operation and an addition done in a for loop. There is no need to muck about with the strides because when rotating 180 degrees the strides are going to be identical. all that is being done is the last pixel is being moved to the first pixel location and vice versa through the entire buffer. 

at a 25% speed increase at a minimum this is something that I feel should be done. This can also be done for the ARGB8888 and RGB888 180° functions as well. The RGB888 function will have some more addition in it.

**Comment by MacDragon** [comment]
The reason is that the function is not intended to only rotate a whole display frame buffer, but insert a smaller drawn canvas rotated into a larger frame buffer.

**Comment by kdschlosser** [comment]
It doesn't matter if you are doing partial updating or not. The strides will always be the same when doing a 180 degree rotation. The buffers must always be the same size in order to do the rotations. 

What would be the use case to do a 180 degree rotation using 2 different sized buffers and having different strides???

**Comment by MacDragon** [comment]
lv_draw_sw_rotate called in e.g. SDL and linux FB flush_cb's in library when LV_DISPLAY_RENDER_MODE_PARTIAL with small input buffer and output location in full FB, thus two different strides.

**Comment by kdschlosser** [comment]
You failed to notice this...

https://github.com/lvgl/lvgl/blob/d824618366e474eef57081d647031965e6edc136/src/drivers/sdl/lv_sdl_window.c#L226

There is a temporary buffer that gets created that is identical in size to the partial buffer and that temporary buffer is where the rotated data gets copied to. As you can see that buffer is the EXACT same size as the buffer that is holding the data. The same thing happens in the linux fb driver.. 

https://github.com/lvgl/lvgl/blob/d824618366e474eef57081d647031965e6edc136/src/drivers/display/fb/lv_linux_fbdev.c#L274


There is nothing that gets broken from the change above. it will work 100% as it should without any issue.

**Comment by MacDragon** [comment]
the lv_area_t describes the output coordinates within the fullsize framebuffer in flush_cb for a partial render, and they match the size of the partial buffer given - that is enough information to work out where to render it.

**Comment by kdschlosser** [comment]
Yeah the area that is passed to the flush callback. But that area is never passed to the rotate functions. The full frame buffer is not passed to the rotation functions either. So the data has to be copied from the partial buffer to a temp buffer and after that is done the data from the temp buffer needs to get copied to the full frame buffer,. why not copy the data from the partial buffer directly to the full frame buffer?

**Comment by MacDragon** [comment]
I mean that it can be done that way with the information given - in our own flusb_cb that we use it does the a rotated copy straight into FB without any additional buffers, I had assumed that they functions were working the same way, hence not looking at them too carefully.

**Comment by kdschlosser** [comment]
Hey if you want to change the code for those rotation functions that will do the rotation and the copy to a larger buffer setting the position of the data properly I am pretty sure you will not see anyone telling you no it won't be accepted. It sure would make things a hell of a lot easier on the user if needing to do a software rotation and not to mention it would also make it a hell of a lot faster too.

**Comment by kisvegabor** [comment]
Probably we could rotate directly into the frame buffers too. Just need to go to to the first pixel of the frame buffer (top left corner of the rotated image) and and use it the destination buffer.

**Comment by kisvegabor** [comment]
> That's where the 3rd much smaller partial buffer comes into play. 

Seems doable. It's something that can be managed in the flush_cb and not modification to LVGL is needed.

> I know that LVGL has the lv_display_flush_is_last.. what is the purpose of this function? Is it the way to determine if all of the updates have been written? If it is this would be the way to identify when to swap the frame buffers buffers

Exactly!

---

I've opened a PR to update SDL: #6835

**Comment by kdschlosser** [comment]
Ummm. Dunno why this got closed because it is not yet finished...

There is also the Linux fb driver that needs to be updated in a similar fashion mas the SDL driver.

**Comment by kisvegabor** [comment]
It was closed because  #6835 was merged. 

@kdschlosser can you confirm that the rotation update that I added for SDL is a  good direction. If so, I can add the same for fbdev and maybe other too.

**Comment by kdschlosser** [comment]
I will check it out a little bit later today. I will run the benchmark using the old code vs the new and see if there is any improvement.

**Comment by kdschlosser** [comment]
I wanted to drop some code in here for folks that might want to perform a copy from a partial buffer into a full buffer while handling rotation at the same time. Optionally performing an RGB565 byte swap can be done at the same time as well as dithering for RGB888 and RGB565. All optional pieces are able to be turned on and off at runtime instead of at compile time.

<details>

<summary>rotation.h</summary>

```
#include <stdint.h>
#include <stddef.h>
#include <stdbool.h>
#include <string.h>


#ifndef __ROTATION_H__
    #define __ROTATION_H__

    #define LCD_ROTATION_0    (0)
    #define LCD_ROTATION_90   (1)
    #define LCD_ROTATION_180  (2)
    #define LCD_ROTATION_270  (3)

    typedef struct _lcd_tx_data_t {
        uint8_t *data;
        int x_start;
        int y_start;
        int x_end;
        int y_end;
        uint32_t rotation: 2;
        uint32_t dither: 1;
        uint32_t last_update: 1;
        uint32_t byte_swap: 1;
        uint32_t sw_rotation: 1;
        uint32_t bytes_per_pixel: 4;
        uint32_t dst_width: 11;
        uint32_t dst_height: 11;
    } lcd_tx_data_t;

    void copy_pixels(void *dst, lcd_tx_data_t *tx_data);

#endif
```

</details>



<details>

<summary>rotation.c</summary>

```c

#include <stdint.h>
#include <stddef.h>
#include <stdbool.h>
#include <string.h>

#include "rotation.h"


#define _CALC_THRESHOLD(x, y)  (uint8_t)(((y & 7) << 3) + (x & 7))


uint8_t* red_thresh = NULL;
uint8_t* green_thresh = NULL;
uint8_t* blue_thresh = NULL;


void dither_deinit(void)
{
    if (red_thresh != NULL) m_free(red_thresh);
    if (green_thresh != NULL) m_free(green_thresh);
    if (blue_thresh != NULL) m_free(blue_thresh);

    red_thresh = NULL;
    green_thresh = NULL;
    blue_thresh = NULL;
    return;
}


bool dither_init(void)
{
    red_thresh = (uint8_t *)m_malloc(64);
    if (red_thresh != NULL) {
        memcpy(red_thresh,
            (uint8_t []){
                1, 7, 3, 5, 0, 8, 2, 6,
                7, 1, 5, 3, 8, 0, 6, 2,
                3, 5, 0, 8, 2, 6, 1, 7,
                5, 3, 8, 0, 6, 2, 7, 1,
                0, 8, 2, 6, 1, 7, 3, 5,
                8, 0, 6, 2, 7, 1, 5, 3,
                2, 6, 1, 7, 3, 5, 0, 8,
                6, 2, 7, 1, 5, 3, 8, 0
        }, 64);
    }

    green_thresh = (uint8_t *)m_malloc(64);
    if (green_thresh != NULL) {
        memcpy(green_thresh,
            (uint8_t []){
                1, 3, 2, 2, 3, 1, 2, 2,
                2, 2, 0, 4, 2, 2, 4, 0,
                3, 1, 2, 2, 1, 3, 2, 2,
                2, 2, 4, 0, 2, 2, 0, 4,
                1, 3, 2, 2, 3, 1, 2, 2,
                2, 2, 0, 4, 2, 2, 4, 0,
                3, 1, 2, 2, 1, 3, 2, 2,
                2, 2, 4, 0, 2, 2, 0, 4
        }, 64);
    }

    blue_thresh = (uint8_t *)m_malloc(64);
    if (blue_thresh != NULL) {
        memcpy(blue_thresh,
            (uint8_t []){
                5, 3, 8, 0, 6, 2, 7, 1,
                3, 5, 0, 8, 2, 6, 1, 7,
                8, 0, 6, 2, 7, 1, 5, 3,
                0, 8, 2, 6, 1, 7, 3, 5,
                6, 2, 7, 1, 5, 3, 8, 0,
                2, 6, 1, 7, 3, 5, 0, 8,
                7, 1, 5, 3, 8, 0, 6, 2,
                1, 7, 3, 5, 0, 8, 2, 6
        }, 64);
    }

    if (red_thresh == NULL || blue_thresh == NULL || green_thresh == NULL) {
        dither_deinit();
        return false;
    } else {
        return true;
    }
}


static inline void dither_swap_16bpp(uint8_t treshold_id, uint16_t *from, uint16_t *to)
{
    *to = (
        ((((*from >> 8) & 0xF8) + red_thresh[treshold_id]) << 8) |
        ((((*from >> 3) & 0xFC) + green_thresh[treshold_id]) << 3) |
        (((((*from++) & 0x1F) << 3) + blue_thresh[treshold_id]) >> 3)
    );
    *to++ = (*to >> 8) | (*to << 8);
}

static inline void swap_16bpp(uint16_t *to, uint16_t *from)
{
    *to++ = (*from >> 8) | ((*from++) << 8);
}


static inline void dither_16bpp(uint8_t treshold_id, uint16_t *from, uint16_t *to)
{
    *to++ = (
        ((((*from >> 8) & 0xF8) + red_thresh[treshold_id]) << 8) |
        ((((*from >> 3) & 0xFC) + green_thresh[treshold_id]) << 3) |
        (((((*from++) & 0x1F) << 3) + blue_thresh[treshold_id]) >> 3)
    );
}


static inline void dither_24bpp(uint8_t treshold_id, uint8_t *from, uint8_t *to)
{
    *to++ = (*from++) + red_thresh[treshold_id];
    *to++ = (*from++) + green_thresh[treshold_id];
    *to++ = (*from++) + blue_thresh[treshold_id];
}


static inline void copy_8bpp(uint8_t *from, uint8_t *to)
{
    *to++ = *from++;
}


static inline void copy_16bpp(uint16_t *from, uint16_t *to)
{
    *to++ = *from++;
}


static inline void copy_24bpp(uint8_t *from, uint8_t *to)
{
    *to++ = *from++;
    *to++ = *from++;
    *to++ = *from++;
}


static inline void copy_32bpp(uint32_t *from, uint32_t *to)
{
    *to++ = *from++;
}


static void rotate0(uint8_t *src, uint8_t *dst, uint32_t x_start, uint32_t y_start,
                    uint32_t x_end, uint32_t y_end, uint32_t dst_width, uint32_t dst_height,
                    uint8_t bytes_per_pixel, uint8_t dither, uint8_t rgb565_byteswap);

static void rotate_8bpp(uint8_t *src, uint8_t *dst, uint32_t x_start, uint32_t y_start,
                        uint32_t x_end, uint32_t y_end, uint32_t dst_width, uint32_t dst_height,
                        uint8_t rotate);

static void rotate_16bpp(uint16_t *src, uint16_t *dst, uint32_t x_start, uint32_t y_start,
                         uint32_t x_end, uint32_t y_end, uint32_t dst_width, uint32_t dst_height,
                         uint8_t rotate, uint8_t dither, uint8_t byteswap);

static void rotate_24bpp(uint8_t *src, uint8_t *dst, uint32_t x_start, uint32_t y_start,
                         uint32_t x_end, uint32_t y_end, uint32_t dst_width, uint32_t dst_height,
                         uint8_t rotate, uint8_t dither);

static void rotate_32bpp(uint32_t *src, uint32_t *dst, uint32_t x_start, uint32_t y_start,
                         uint32_t x_end, uint32_t y_end, uint32_t dst_width, uint32_t dst_height,
                         uint8_t rotate);


void copy_pixels(void *dst, lcd_tx_data_t *tx_data)
{
    uint32_t x_start         = (uint32_t)tx_data->x_start;
    uint32_t y_start         = (uint32_t)tx_data->y_start;
    uint32_t x_end           = (uint32_t)tx_data->x_end;
    uint32_t y_end           = (uint32_t)tx_data->y_end;
    uint32_t dst_width       = (uint32_t)tx_data->dst_width;
    uint32_t dst_height      = (uint32_t)tx_data->dst_height;
    uint8_t  bytes_per_pixel = (uint8_t )tx_data->bytes_per_pixel;
    uint8_t  rotation        = (uint8_t )tx_data->rotation;
    
    if (bytes_per_pixel == 2 || bytes_per_pixel == 3) {
        if (tx_data->dither && red_thresh == NULL && !dither_init()) tx_data->dither = 0;
        else if (!tx_data->dither && red_thresh != NULL) dither_deinit();
    }
    
    if (rotation == LCD_ROTATION_0) {
        x_start = MIN(x_start, dst_width );
        y_start = MIN(y_start, dst_height);
        x_end   = MIN(x_end,   dst_width );
        y_end   = MIN(y_end,   dst_height);
        
        rotate0((void*)tx_data->data, dst, 
                x_start, y_start, 
                x_end, y_end, 
                dst_width, dst_height, 
                tx_data->dither, tx_data->byte_swap, bytes_per_pixel);
    } else {
        y_end += 1; // removes black lines between blocks
        
        if (rotation == LCD_ROTATION_90 || rotation == LCD_ROTATION_270) {
            x_start = MIN(x_start, tx_data->dst_height);
            x_end   = MIN(x_end,   tx_data->dst_height);
            y_start = MIN(y_start, tx_data->dst_width );
            y_end   = MIN(y_end,   tx_data->dst_width );
            
        } else {
            x_start = MIN(x_start, dst_width );
            x_end   = MIN(x_end,   dst_width );
            y_start = MIN(y_start, dst_height);
            y_end   = MIN(y_end,   dst_height);
        }
        
        switch(bytes_per_pixel) {
            case 1: 
                rotate_8bpp((void*)tx_data->data, dst, x_start, y_start, 
                            x_end, y_end, dst_width, dst_height, rotation);
                break;
            case 2:
                rotate_16bpp((void*)tx_data->data, dst, x_start, y_start, 
                             x_end, y_end, dst_width, dst_height, rotation, 
                             tx_data->dither, tx_data->byte_swap);
                break;
            case 3:
                rotate_24bpp((void*)tx_data->data, dst, x_start, y_start, 
                             x_end, y_end, dst_width, dst_height, rotation, 
                             tx_data->dither);
                break;
            case 4:
                rotate_32bpp((void*)tx_data->data, dst, x_start, y_start, 
                             x_end, y_end, dst_width, dst_height, rotation);
                break
            default:
                break;
        }
    }
}

void rotate0(uint8_t *src,       uint8_t *dst, 
             uint32_t x_start,   uint32_t y_start,
             uint32_t x_end,     uint32_t y_end, 
             uint32_t dst_width, uint32_t dst_height, 
             uint8_t dither,     uint8_t byteswap,    
             uint8_t bytes_per_pixel)
{
    dst += ((y_start * dst_width + x_start) * bytes_per_pixel);
    
    if(x_start == 0 && x_end == (dst_width - 1) && !dither && !byteswap) {
        memcpy(dst, src, dst_width * (y_end - y_start + 1) * bytes_per_pixel);
    } else {
        uint32_t src_bytes_per_line = (x_end - x_start + 1) * bytes_per_pixel;
        uint32_t dst_bytes_per_line = dst_width * bytes_per_pixel;

        switch(bytes_per_pixel) {
            case 2:
                if (dither && byteawap) {
                    for(uint32_t y = y_start; y < y_end; y++) {
                    
                        for (uint32_t x=0;x<x_end;x++) {
                            dither_swap_16bpp(_CALC_THRESHOLD(x, y), 
                                              (uint16_t *)(src) + x, 
                                              (uint16_t *)(dst) + x);
                        }
                        dst += dst_bytes_per_line;
                        src += src_bytes_per_line;
                    }
                    return;
                } else if (byteawap) {
                    for(uint32_t y = y_start; y < y_end; y++) {
                    
                        for (uint32_t x=0;x<x_end;x++) {
                            swap_16bpp((uint16_t *)(src) + x, 
                                       (uint16_t *)(dst) + x);
                        }
                        dst += dst_bytes_per_line;
                        src += src_bytes_per_line;
                    }
                    return;
                } else if (dither) {
                    for(uint32_t y = y_start; y < y_end; y++) {
                    
                        for (uint32_t x=0;x<x_end;x++) {
                            dither_16bpp(_CALC_THRESHOLD(x, y), 
                                         (uint16_t *)(src) + x, 
                                         (uint16_t *)(dst) + x);
                        }
                        dst += dst_bytes_per_line;
                        src += src_bytes_per_line;
                    }
                    return;
                }
            case 3:
                if (dither) {
                    for(uint32_t y = y_start; y < y_end; y++) {
                    
                        for (uint32_t x=0;x<x_end;x++) {
                            dither_24bpp(_CALC_THRESHOLD(x, y), 
                                         (uint8_t *)(src) + (x * 3), 
                                         (uint8_t *)(dst) + (x * 3));
                        }
                        dst += dst_bytes_per_line;
                        src += src_bytes_per_line;
                    }
                    return;
                }
            default:
                break;
        }

        for(uint32_t y = y_start; y < y_end; y++) {
            memcpy(dst, src, src_bytes_per_line);
            dst += dst_bytes_per_line;
            src += src_bytes_per_line;
        }
    }
}


void rotate_8bpp(uint8_t *src,       uint8_t *dst, 
                 uint32_t x_start,   uint32_t y_start,
                 uint32_t x_end,     uint32_t y_end, 
                 uint32_t dst_width, uint32_t dst_height,
                 uint8_t rotate)
{
    uint32_t i;
    uint32_t j;

    uint32_t src_bytes_per_line = x_end - x_start + 1;
    uint32_t offset = y_start * src_bytes_per_line + x_start;

    switch (rotate) {
        case LCD_ROTATION_90:
            dst_height--;

            for (uint32_t y = y_start; y < y_end; y++) {
                i = y * src_bytes_per_line;
                j = dst_width + y;

                for (uint32_t x = x_start; x < x_end; x++) {
                    copy_8bpp(src + (i + x - offset),
                              dst + ((dst_height - x) * j));
                }
            }
            break;

        // MIRROR_X MIRROR_Y
        case LCD_ROTATION_180:
            LCD_UNUSED(src_bytes_per_line);
            LCD_UNUSED(offset);

            j = dst_width - 1 - x_start;
            dst_height--;

            for (uint32_t y = y_start; y < y_end; y++) {
                i = ((dst_height - y) * dst_width) + j;

                for (uint32_t x = x_start; x < x_end; x++) {
                    copy_8bpp(src, dst + i);
                    src++;
                    i--;
                }
                src++;
            }
            break;

        // SWAP_XY   MIRROR_X
        case LCD_ROTATION_270:
            uint32_t dst_width_minus_one = dst_width - 1;

            for (uint32_t y = y_start; y < y_end; y++) {
                j = dst_width_minus_one - y;
                i = y * src_bytes_per_line;
                
                for (uint32_t x = x_start; x < x_end; x++) {
                    copy_8bpp(src + (i + x - offset),
                              dst + ((x * dst_width) + j));
                }
            }
            break;

        default:
            LCD_UNUSED(i);
            LCD_UNUSED(j);
            LCD_UNUSED(src_bytes_per_line);
            LCD_UNUSED(offset);
            break;
    }

}


void rotate_16bpp(uint16_t *src,      uint16_t *dst, 
                  uint32_t x_start,   uint32_t y_start,
                  uint32_t x_end,     uint32_t y_end, 
                  uint32_t dst_width, uint32_t dst_height,
                  uint8_t rotate,     uint8_t dither,      
                  uint8_t byteswap)
{
    uint32_t i;
    uint32_t j;

    uint32_t src_bytes_per_line = x_end - x_start + 1;
    uint32_t offset = y_start * src_bytes_per_line + x_start;
    uint32_t dst_width_minus_one = 0;

    if (dither && byteswap) {
        switch (rotate) {
            case LCD_ROTATION_90:
                LCD_UNUSED(dst_width_minus_one);
                dst_height--;

                for (uint32_t y = y_start; y < y_end; y++) {
                    i = y * src_bytes_per_line;
                    j = dst_width + y;

                    for (uint32_t x = x_start; x < x_end; x++) {
                        dither_swap_16bpp(_CALC_THRESHOLD(x, y),
                                          src + (i + x - offset),
                                          dst + ((dst_height - x) * j));
                    }
                }
                break;

            // MIRROR_X MIRROR_Y
            case LCD_ROTATION_180:
                LCD_UNUSED(dst_width_minus_one);
                LCD_UNUSED(src_bytes_per_line);
                LCD_UNUSED(offset);

                j = dst_width - 1 - x_start;
                dst_height--;

                for (uint32_t y = y_start; y < y_end; y++) {
                    i = ((dst_height - y) * dst_width) + j;

                    for (uint32_t x = x_start; x < x_end; x++) {
                        dither_swap_16bpp(_CALC_THRESHOLD(x, y), src, dst + i);
                        src++;
                        i--;
                    }
                    src++;
                }
                break;

            // SWAP_XY   MIRROR_X
            case LCD_ROTATION_270:
                dst_width_minus_one = dst_width - 1;

                for (uint32_t y = y_start; y < y_end; y++) {
                    j = dst_width_minus_one - y;
                    i = y * src_bytes_per_line;

                    for (uint32_t x = x_start; x < x_end; x++) {
                        i = y * src_bytes_per_line + x - offset;
                        j = (x * dst_width + dst_width - 1 - y);
                        dither_swap_16bpp(_CALC_THRESHOLD(x, y),
                                          src + (i + x - offset),
                                          dst + ((x * dst_width) + j));
                    }
                }
                break;

            default:
                LCD_UNUSED(dst_width_minus_one);
                LCD_UNUSED(i);
                LCD_UNUSED(j);
                LCD_UNUSED(src_bytes_per_line);
                LCD_UNUSED(offset);
                break;
        }
    } else if (dither) {
        switch (rotate) {
            case LCD_ROTATION_90:
                LCD_UNUSED(dst_width_minus_one);

                dst_height--;

                for (uint32_t y = y_start; y < y_end; y++) {
                    i = y * src_bytes_per_line;
                    j = dst_width + y;

                    for (uint32_t x = x_start; x < x_end; x++) {
                        dither_16bpp(_CALC_THRESHOLD(x, y),
                                     src + (i + x - offset),
                                     dst + ((dst_height - x) * j));
                    }
                }
                break;

            // MIRROR_X MIRROR_Y
            case LCD_ROTATION_180:
                LCD_UNUSED(dst_width_minus_one);
                LCD_UNUSED(src_bytes_per_line);
                LCD_UNUSED(offset);

                j = dst_width - 1 - x_start;
                dst_height--;

                for (uint32_t y = y_start; y < y_end; y++) {
                    i = ((dst_height - y) * dst_width) + j;

                    for (uint32_t x = x_start; x < x_end; x++) {
                        dither_16bpp(_CALC_THRESHOLD(x, y), src, dst + i);
                        src++;
                        i--;
                    }
                    src++;
                }
                break;

            // SWAP_XY   MIRROR_X
            case LCD_ROTATION_270:
                dst_width_minus_one = dst_width - 1;

                for (uint32_t y = y_start; y < y_end; y++) {
                    j = dst_width_minus_one - y;
                    i = y * src_bytes_per_line;

                    for (uint32_t x = x_start; x < x_end; x++) {
                        dither_16bpp(
                            _CALC_THRESHOLD(x, y),
                            src + (i + x - offset),
                            dst + ((x * dst_width) + j)
                        );
                    }
                }
                break;

            default:
                LCD_UNUSED(dst_width_minus_one);
                LCD_UNUSED(i);
                LCD_UNUSED(j);
                LCD_UNUSED(src_bytes_per_line);
                LCD_UNUSED(offset);
                break;
        }

    } else if (byteswap) {
        switch (rotate) {
            case LCD_ROTATION_90:
                LCD_UNUSED(dst_width_minus_one);

                dst_height--;

                for (uint32_t y = y_start; y < y_end; y++) {
                    i = y * src_bytes_per_line;
                    j = dst_width + y;

                    for (uint32_t x = x_start; x < x_end; x++) {
                        swap_16bpp(src + (i + x - offset),
                                   dst + ((dst_height - x) * j));
                    }
                }
                break;

            // MIRROR_X MIRROR_Y
            case LCD_ROTATION_180:
                LCD_UNUSED(dst_width_minus_one);
                LCD_UNUSED(src_bytes_per_line);
                LCD_UNUSED(offset);

                j = dst_width - 1 - x_start;
                dst_height--;

                for (uint32_t y = y_start; y < y_end; y++) {
                    i = ((dst_height - y) * dst_width) + j;

                    for (uint32_t x = x_start; x < x_end; x++) {
                        swap_16bpp(src, dst + i);
                        src++;
                        i--;
                    }
                    src++;
                }
                break;

            // SWAP_XY   MIRROR_X
            case LCD_ROTATION_270:
                dst_width_minus_one = dst_width - 1;

                for (uint32_t y = y_start; y < y_end; y++) {
                    j = dst_width_minus_one - y;
                    i = y * src_bytes_per_line;

                    for (uint32_t x = x_start; x < x_end; x++) {
                        swap_16bpp(src + (i + x - offset),
                                   dst + ((x * dst_width) + j));
                    }
                }
                break;

            default:
                LCD_UNUSED(dst_width_minus_one);
                LCD_UNUSED(i);
                LCD_UNUSED(j);
                LCD_UNUSED(src_bytes_per_line);
                LCD_UNUSED(offset);
                break;
        }
    } else {
        switch (rotate) {
            case LCD_ROTATION_90:
                LCD_UNUSED(dst_width_minus_one);

                dst_height--;

                for (uint32_t y = y_start; y < y_end; y++) {
                    i = y * src_bytes_per_line;
                    j = dst_width + y;

                    for (uint32_t x = x_start; x < x_end; x++) {
                        copy_16bpp(src + (i + x - offset),
                                   dst + ((dst_height - x) * j));
                    }
                }
                break;

            // MIRROR_X MIRROR_Y
            case LCD_ROTATION_180:
                LCD_UNUSED(dst_width_minus_one);
                LCD_UNUSED(src_bytes_per_line);
                LCD_UNUSED(offset);

                j = dst_width - 1 - x_start;
                dst_height--;

                for (uint32_t y = y_start; y < y_end; y++) {
                    i = ((dst_height - y) * dst_width) + j;

                    for (uint32_t x = x_start; x < x_end; x++) {
                        copy_16bpp(src, dst + i);
                        src++;
                        i--;
                    }
                    src++;
                }
                break;

            // SWAP_XY   MIRROR_X
            case LCD_ROTATION_270:
                dst_width_minus_one = dst_width - 1;

                for (uint32_t y = y_start; y < y_end; y++) {
                    j = dst_width_minus_one - y;
                    i = y * src_bytes_per_line;

                    for (uint32_t x = x_start; x < x_end; x++) {
                        copy_16bpp(src + (i + x - offset),
                                   dst + ((x * dst_width) + j));
                    }
                }
                break;

            default:
                LCD_UNUSED(dst_width_minus_one);
                LCD_UNUSED(i);
                LCD_UNUSED(j);
                LCD_UNUSED(src_bytes_per_line);
                LCD_UNUSED(offset);
                break;
        }
    }
}


void rotate_24bpp(uint8_t *src,       uint8_t *dst, 
                  uint32_t x_start,   uint32_t y_start,
                  uint32_t x_end,     uint32_t y_end, 
                  uint32_t dst_width, uint32_t dst_height,
                  uint8_t rotate,     uint8_t dither)
{
    uint32_t i;
    uint32_t j;

    uint32_t src_bytes_per_line = (x_end - x_start + 1) * 3;
    uint32_t offset = y_start * src_bytes_per_line + x_start * 3;
    uint32_t dst_width_minus_one = 0;

    if (dither) {
        switch (rotate) {
            case LCD_ROTATION_90:
                LCD_UNUSED(dst_width_minus_one);

                dst_height--;

                for (uint32_t y = y_start; y < y_end; y++) {
                    i = y * src_bytes_per_line;
                    j = dst_width + y;

                    for (uint32_t x = x_start; x < x_end; x++) {
                        dither_24bpp(_CALC_THRESHOLD(x, y),
                                     src + (i + (x * 3) - offset),
                                     dst + ((dst_height - x) * j * 3));
                    }
                }
                break;

            // MIRROR_X MIRROR_Y
            case LCD_ROTATION_180:
                LCD_UNUSED(dst_width_minus_one);
                LCD_UNUSED(src_bytes_per_line);
                LCD_UNUSED(offset);

                j = dst_width - 1 - x_start;
                dst_height--;

                for (uint32_t y = y_start; y < y_end; y++) {
                    i = (((dst_height - y) * dst_width) + j) * 3;

                    for (uint32_t x = x_start; x < x_end; x++) {
                        dither_24bpp(_CALC_THRESHOLD(x, y), src, dst + i);
                        src += 3;
                        i -= 3;
                    }
                    src++;
                }
                break;

            // SWAP_XY   MIRROR_X
            case LCD_ROTATION_270:
                dst_width_minus_one = dst_width - 1;

                for (uint32_t y = y_start; y < y_end; y++) {
                    j = dst_width_minus_one - y;
                    i = y * src_bytes_per_line;

                    for (uint32_t x = x_start; x < x_end; x++) {
                        dither_24bpp(_CALC_THRESHOLD(x, y),
                                     src + (i + (x * 3) - offset),
                                     dst + (((x * dst_width) + j) * 3));
                    }
                }
                break;

            default:
                LCD_UNUSED(dst_width_minus_one);
                LCD_UNUSED(i);
                LCD_UNUSED(j);
                LCD_UNUSED(src_bytes_per_line);
                LCD_UNUSED(offset);
                break;
        }
    } else {
        switch (rotate) {
            case LCD_ROTATION_90:
                LCD_UNUSED(dst_width_minus_one);
                dst_height--;

                for (uint32_t y = y_start; y < y_end; y++) {
                    i = y * src_bytes_per_line;
                    j = dst_width + y;

                    for (uint32_t x = x_start; x < x_end; x++) {
                        copy_24bpp(src + (i + (x * 3) - offset),
                                   dst + ((dst_height - x) * j * 3));
                    }
                }
                break;

            // MIRROR_X MIRROR_Y
            case LCD_ROTATION_180:
                LCD_UNUSED(dst_width_minus_one);
                LCD_UNUSED(src_bytes_per_line);
                LCD_UNUSED(offset);

                j = dst_width - 1 - x_start;
                dst_height--;

                for (uint32_t y = y_start; y < y_end; y++) {
                    i = (((dst_height - y) * dst_width) + j) * 3;

                    for (size_t x = x_start; x < x_end; x++) {
                        copy_24bpp(src, dst + i);
                        src += 3;
                        i -= 3;
                    }
                    src++;
                }
                break;

            // SWAP_XY   MIRROR_X
            case LCD_ROTATION_270:
                dst_width_minus_one = dst_width - 1;

                for (uint32_t y = y_start; y < y_end; y++) {
                    j = dst_width_minus_one - y;
                    i = y * src_bytes_per_line;

                    for (uint32_t x = x_start; x < x_end; x++) {
                        copy_24bpp(src + (i + (x * 3) - offset),
                                   dst + (((x * dst_width) + j) * 3));
                    }
                }
                break;

            default:
                LCD_UNUSED(dst_width_minus_one);
                LCD_UNUSED(i);
                LCD_UNUSED(j);
                LCD_UNUSED(src_bytes_per_line);
                LCD_UNUSED(offset);
                break;
        }
    }
}


void rotate_32bpp(uint32_t *src,      uint32_t *dst, 
                  uint32_t x_start,   uint32_t y_start,
                  uint32_t x_end,     uint32_t y_end, 
                  uint32_t dst_width, uint32_t dst_height,
                  uint8_t rotate)
{
    uint32_t i;
    uint32_t j;

    uint32_t src_bytes_per_line = x_end - x_start + 1;
    uint32_t offset = y_start * src_bytes_per_line + x_start;

    switch (rotate) {
        case LCD_ROTATION_90:
            dst_height--;

            for (uint32_t y = y_start; y < y_end; y++) {
                i = y * src_bytes_per_line;
                j = dst_width + y;

                for (uint32_t x = x_start; x < x_end; x++) {
                    copy_32bpp(src + (i + x - offset),
                               dst + ((dst_height - x) * j));
                }
            }
            break;

        // MIRROR_X MIRROR_Y
        case LCD_ROTATION_180:
            LCD_UNUSED(src_bytes_per_line);
            LCD_UNUSED(offset);

            j = dst_width - 1 - x_start;
            dst_height--;

            for (uint32_t y = y_start; y < y_end; y++) {
                i = ((dst_height - y) * dst_width) + j;

                for (uint32_t x = x_start; x < x_end; x++) {
                    copy_32bpp(src, dst + i);
                    src++;
                    i--;
                }
                src++;
            }
            break;

        // SWAP_XY   MIRROR_X
        case LCD_ROTATION_270:
            uint32_t dst_width_minus_one = dst_width - 1;

            for (uint32_t y = y_start; y < y_end; y++) {
                j = dst_width_minus_one - y;
                i = y * src_bytes_per_line;
                
                for (uint32_t x = x_start; x < x_end; x++) {
                    copy_32bpp(src + (i + x - offset),
                               dst + ((x * dst_width) + j));
                }
            }
            break;

        default:
            LCD_UNUSED(i);
            LCD_UNUSED(j);
            LCD_UNUSED(src_bytes_per_line);
            LCD_UNUSED(offset);
            break;
    }
}
```


</details>
